zcash_client_sqlite/src/transact.rs

```diff
@@ -0,0 +1,848 @@
+            };
+
+            let from = extfvk
+                .fvk
+                .vk
+                .to_payment_address(diversifier, &JUBJUB)
+                .unwrap();
+            let note = from.create_note(note_value as u64, rcm, &JUBJUB).unwrap();
+
+            let merkle_path = {
+                let d: Vec<_> = row.get(3)?;
+                IncrementalWitness::read(&d[..])?
+                    .path()
+                    .expect("the tree is not empty")
+            };
+
+            Ok(SelectedNoteRow {
+                diversifier,
+                note,
+                merkle_path,
+            })
+        },
+    )?;
+    let notes: Vec<SelectedNoteRow> = notes.collect::<Result<_, _>>()?;
+
+    // Confirm we were able to select sufficient value
+    let selected_value = notes
+        .iter()
+        .fold(0, |acc, selected| acc + selected.note.value);
+    if selected_value < target_value as u64 {
+        return Err(Error(ErrorKind::InsufficientBalance(
+            selected_value,
+            target_value as u64,
+        )));
+    }
+
+    // Create the transaction
+    let mut builder = Builder::new(height);
+    for selected in notes {
+        builder.add_sapling_spend(
+            extsk.clone(),
+            selected.diversifier,
+            selected.note,
+            selected.merkle_path,
+        )?;
+    }
+    match to {
+        RecipientAddress::Shielded(to) => {
+            builder.add_sapling_output(ovk, to.clone(), value, memo.clone())
+        }
+        RecipientAddress::Transparent(to) => builder.add_transparent_output(&to, value),
+    }?;
+    let (tx, tx_metadata) = builder.build(consensus_branch_id, &prover)?;
+    // We only called add_sapling_output() once.
+    let output_index = match tx_metadata.output_index(0) {
+        Some(idx) => idx as i64,
+        None => panic!("Output 0 should exist in the transaction"),
+    };
+    let created = time::get_time();
+
+    // Update the database atomically, to ensure the result is internally consistent.
+    data.execute("BEGIN IMMEDIATE", NO_PARAMS)?;
+
+    // Save the transaction in the database.
+    let mut raw_tx = vec![];
+    tx.write(&mut raw_tx)?;
+    let mut stmt_insert_tx = data.prepare(
+        "INSERT INTO transactions (txid, created, expiry_height, raw)
+        VALUES (?, ?, ?, ?)",
+    )?;
+    stmt_insert_tx.execute(&[
+        tx.txid().0.to_sql()?,
+        created.to_sql()?,
+        tx.expiry_height.to_sql()?,
+        raw_tx.to_sql()?,
+    ])?;
+    let id_tx = data.last_insert_rowid();
+
+    // Mark notes as spent.
+    //
+    // This locks the notes so they aren't selected again by a subsequent call to
+    // create_to_address() before this transaction has been mined (at which point the notes
+    // get re-marked as spent).
+    //
+    // Assumes that create_to_address() will never be called in parallel, which is a
+    // reasonable assumption for a light client such as a mobile phone.
```

Comment:
This assumption should be documented in the function comment.

Reply:
It is:
> Do not call this multiple times in parallel, or you will generate transactions that double-spend the same notes.